src/app/scheme/scheme.service.ts

```diff
@@ -432,63 +458,222 @@ export class SchemeService extends ISchemeService {
+        return this.http.patch<Patch_Structure_Response<T>>(url, dataToUpdate)
+            .pipe(tap((response) => {
+                if (!response.result) return;
+
+                const findObj = (array: Array<T>, item: T, ignoreKeys = ['id']) => {
+                    return array.find((i) => {
+                        const keys = this.diff(item, i);
+                        return keys.filter(key => ignoreKeys.indexOf(key as string) === -1).length === 0;
+                    }) as T;
+                };
+
+                const findGroup = (id: number) => {
+                    return this.scheme.section.map(sct => sct.groups)
+                        .reduce((prev, curr) => {
+                            prev.push(...curr);
+                            return prev;
+                        }, [])
+                        .find(group => id === group.id);
+                };
+
+                const findAndUpdate = (array: Array<T>, item: T): boolean => {
+                    const itemToUpdate = array.find(({ id }) => item.id === id);
+                    if (!itemToUpdate) {
+                        return false;
+                    }
+
+                    Object.assign(itemToUpdate, item);
+                    return true;
+                };
+
+                const findAndRemove = (array: Array<T>, item: T) => {
+                    const itemIdx = array.findIndex(({ id }) => item.id === id);
+                    if (itemIdx >= 0) {
+                        array.splice(itemIdx, 1);
+                    }
+                };
+
+                const findDevice = (device_id: number) => {
+                    return this.scheme.device.find(dev => dev.id === device_id);
+                };
+
+                const findSection = (section_id: number) => {
+                    return this.scheme.section.find(sct => sct.id === section_id);
+                };
+
+                data.forEach((changeInfo) => {
+                    let array: Array<T>;
+                    let array2: Array<T>;
+
+                    const { obj, prev, state } = changeInfo;
+
+                    if (settingName === Structure_Type.ST_DEVICE_ITEM_GROUP) {
+                        array = findSection((<Device_Item_Group>obj).section_id)?.groups as any;
+                    } else if (settingName === Structure_Type.ST_DEVICE_ITEM) {
+                        const devItem = obj as Device_Item;
+                        array = findGroup(devItem.group_id).items as any;
+                        array2 = findDevice(devItem.device_id).items as any;
+                    } else if (settingName === Structure_Type.ST_DIG_PARAM) {
+                        array = findGroup((<Omit<DIG_Param, 'value'>>obj).group_id).params as any;
+                    } else if (settingName === Structure_Type.ST_TRANSLATION) {
+                        return; // TODO: fix
+                    } else {
+                        const key = SchemeService.getSchemeFieldKey(settingName);
```

Comment:
Разве это не одно и тоже?
```suggestion
                        const key = settingName;
```